### DISCUSSION TITLE
람다와 관련된 코드 스니펫 해결하기

### DISCUSSION THREAD

**Discussion by corock** [Study-2-Modern-Java-In-Action/Study-2-Modern-Java-in-Action#28] [CHAPTER 3 람다 표현식 - Q&A] [answered]
1. 오류의 원인은 무엇인가?
2. 해결 방안은 어떤 상황에 활용될까?

```java
public class InventoryService {

    public static void subtract(int stock) {
        stock--;

        Runnable r = () -> {
            // 아래 코드에서 컴파일 오류 발생
            System.out.println("현재 재고: " + stock);
        };

        stock *= 2;

        r.run();
    }

}
```

**Comment by Irisation23** [comment, marked as answer]
1. p113 의 내용을 참조 하면, `지역 변수의 제약`을 볼 수 있음.
- 지역 변수는 명시적으로 final로 선언되어 있어야 하거나 실질적으로 `final`로 선언된 변수와 똑같이 사용 되어야 함.

> 따라서 오류의 원인은 `stock` 변수에 값이 두 번 할당 된것이 원인임.
---

2. 람다식에서 사용하고 있는 값을 복사해서 사용할 수 있음.

```java
        final int currentStock = stock;
        Runnable r = () -> {
            System.out.println("현재 재고: " + currentStock);
        };
```

이와 같은 이유는 바로 불변성을 지키기 위함이라 생각함.
> 참조 해보면 좋은 글 - https://github.com/orgs/Study-2-Effective-Java/discussions/37

**Comment by JoisFe** [comment]
IDE는 강력하군요
![image](https://github.com/Study-2-Modern-Java-In-Action/Study-2-Modern-Java-in-Action/assets/90208100/ca6b2f4b-0acb-4a1d-bdfb-307af06af6aa)
- 람다 표현식 내에서 사용되는 변수는 `final` 혹은 `effectively final`이어야함.
- final -> 불변이 되어야 하는구나.. 는 이해하겠지만
-  effectively final???
  - 알아보니 실제로는 final로 선언되지 않았지만, 람다 표현식 내에서 값이 변경되지 않는 변수를 의미라고 합니다.

``` java
 stock--;

        int finalStock = stock;
        Runnable r = () -> {
            // 아래 코드에서 컴파일 오류 발생
            System.out.println("현재 재고: " + finalStock);
        };
```
- finalStock은 stock을 할당받은 후 변경되지 않는 변수이므로 -> effectively final 

@bunsung92 님께서 말씀 해주셨듯 `final로 선언되어 있거나 실질적으로 final로 선언된 변수와 똑같이 사용 되어야 한다`
에서 `실질적으로 final로 선언된 변수와 똑같이 사용` -> `effectively final`이 아닐까 싶습니다.

또한 해당 질문의 내용이 순수 함수와도 관련이 있는 것 같아 참고 디스커션 댓글과 관련이 있지 않나 싶어 참고 링크를 남깁니다. [참고 링크](https://github.com/orgs/Study-2-Effective-Java/discussions/119)
